### PR TITLE
support pass async to child threads

### PR DESCRIPTION
**Pull request by zhaozg**
try to solve #182 and #184

1. use boxporinter will make pass userdata to thread easy.
2. luv_handle_t add extra field to hold application data.

and any advice are welcome.

### CODE REVIEW

**Comment by zhaozg** [comment]
Ping @creationix @rphillips .

This PR failed on travis CI when do PROCESS_CLEANUP_TEST=1 testing, and I can't finger out reason,  If you have time, please give me some advice to pass test.

**Comment by zhaozg** [comment]
I changed userdata new and gc flow, are this change lead fail?

**Comment by rphillips** [comment]
@zhaozg i believe there is a `stop` being called twice on a handle.

**Comment by rphillips** [comment]
might be a double close as well.

**Comment by tarruda** [comment]
Valgrind reports a double-free in the failed test, here's the first part of the log:

```
==19935== memcheck, a memory error detector
==19935== copyright (c) 2002-2013, and gnu gpl'd, by julian seward et al.
==19935== using valgrind-3.10.0.svn and libvex; rerun with -h for copyright info
==19935== command: ./build/lua examples/talking-to-children.lua
==19935== parent pid: 19210
==19935== 
==19935== invalid read of size 4
==19935==    at 0x5b2a3f5: uv_walk (in /home/tarruda/dev/luv/build/luv.so)
==19935==    by 0x5b26caa: loop_gc (in /home/tarruda/dev/luv/build/luv.so)
==19935==    by 0x40f131: luad_precall (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x40f5d5: luad_call (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x412ca3: dothecall (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x40e762: luad_rawrunprotected (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x40fd90: luad_pcall (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x412df3: gctm (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x412f40: callallpendingfinalizers (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x413289: luac_freeallobjects (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x41bcef: close_state (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x41c249: lua_close (in /home/tarruda/dev/luv/build/lua)
==19935==  address 0x5742f48 is 88 bytes inside a block of size 312 free'd
==19935==    at 0x4c2bdec: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==19935==    by 0x5b15676: luv_handle_gc (in /home/tarruda/dev/luv/build/luv.so)
==19935==    by 0x40f131: luad_precall (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x40f5d5: luad_call (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x412ca3: dothecall (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x40e762: luad_rawrunprotected (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x40fd90: luad_pcall (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x412df3: gctm (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x412f40: callallpendingfinalizers (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x413289: luac_freeallobjects (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x41bcef: close_state (in /home/tarruda/dev/luv/build/lua)
==19935==    by 0x41c249: lua_close (in /home/tarruda/dev/luv/build/lua)

```

Removing the _gc metamethod from the handle metatable seems to fix it:

```diff
diff --git a/src/luv.c b/src/luv.c
index c4d5041..0042007 100644
--- a/src/luv.c
+++ b/src/luv.c
@@ -406,8 +406,6 @@ static void luv_handle_init(lua_State* L) {
     luaL_newmetatable (L, "uv_"#lc);           \
     lua_pushcfunction(L, luv_handle_tostring); \
     lua_setfield(L, -2, "__tostring");         \
-    lua_pushcfunction(L, luv_handle_gc);       \
-    lua_setfield(L, -2, "__gc");               \
     luaL_newlib(L, luv_##lc##_methods);        \
     luaL_setfuncs(L, luv_handle_methods, 0);   \
     lua_setfield(L, -2, "__index");            \
```

File: src/handle.c [hunk 1]
```diff
@@ -16,10 +16,29 @@
...
+  return handle;
+}
+
+static void* luv_checkudata(lua_State* L, int ud, const char* tname) {
+  return *(void**) luaL_checkudata(L, ud, tname);
+}
+
+static int luv_handle_gc(lua_State* L) {
+  void** udata = lua_touserdata(L, 1);
+  free(*udata);
```
Comment (tarruda) [inline src/handle.c]: This is a problem if the userdata is referenced by the uv_loop_t structure, you need to ensure the handle is closed with `uv_close`, but even so you can't free it immediately without running the loop at least once.

**Comment by zhaozg** [comment]
thanks @tarruda  advice.

**Comment by rphillips** [comment]
+1

**Pull request merged**